pathfinding: capacity-dependent apriori model probability

bitromortac opened the pull request:
## Change Description
Adds a limit for the pathfinding probability, namely that if the amount reaches the capacity of a channel, the success probability should decrease drastically. Fixes #5988 for the current pathfinding system.

To achieve this, we multiply the previous probability with a factor to take capacity into account:

`P *= 1 - 1 / [1 + exp(-(amount - cutoffFactor*capacity)/(smearingFactor*capacity))]`
[graph for cap=1Msat, cutoff=0.75\*cap, smearing=0.1\*cap](https://www.wolframalpha.com/input?i=Plot%5B1-1%2F%281%2Bexp%5B-%28x-750000%29%2F%28100000%29%5D%29%2C+%7Bx%2C0%2C1000000%7D%5D)

This function has the effect that for small amounts we don't alter the current behavior, but for large amounts the probability is reduced. We still consider low-capacity channels and don't throw them away like a hard capacity reduction would do.

This PR introduces building blocks for an alternative description of a probability, see #6815.

### Performance considerations:
I investigated the runtime for the capacity fetching and computation. This is interesting if we want to remove the capacity as a parameter of `Estimator.getPairProbability` and insert a `routingGraph` as a dependency instead, but this seems to add quite some latency. 

* without extra calculation (without last commit):
[DBG] CRTR: Pathfinding perf metrics: nodes=119, edges=10133, time=108.88252ms
* with extra calculation (last commit):
[DBG] CRTR: Pathfinding perf metrics: nodes=250, edges=18004, time=13.57969395s

### Questions:
* The cutoff and scaling parameters need discussion, also whether we want to expose them.
* Capacity computation should be given another thought.

Todo:
- [x] fix itests

joostjager commented on lnrpc/routerrpc/router_backend.go:
```diff
@@ -354,15 +363,22 @@ func (r *RouterBackend) QueryRoutes(ctx context.Context,
 	for _, hop := range rt.Hops {
 		toNode := hop.PubKeyBytes
 
+		capacity, err := r.FetchPairCapacity(fromNode, toNode)
```
Such a dramatic slowdown for an in-memory operation. Did you pull a cpu profile to see what's happening?

joostjager replied:
Interesting. So probably fixable by making a specialized capacity lookup on `GraphCache`? Don't know if it is worth it. It is better separation of concerns though.

bitromortac replied:
Yes, I'll try to see if we can have something performant there.

bitromortac replied:
we decided to go with the uniform policies approach as it avoids a second traversal of a node's peers

lightninglabs-deploy commented:
@bitromortac, remember to re-request review from reviewers when ready

bitromortac commented:
Added a few more unit test cases, fixed a failing itest and made only the apriori hop probability capacity dependent.

joostjager commented on routing/probability_estimator.go:
```diff
@@ -50,18 +66,23 @@ func (p *probabilityEstimator) setConfig(cfg *ProbabilityEstimatorCfg) {
+
+	// We reduce the apriori hop probability if the amount comes close to
+	// the capacity.
+	apriori := p.AprioriHopProbability * capacityFactor(amt, capacity)
```
👍

joostjager commented on routing/probability.go:
```diff
@@ -0,0 +1,23 @@
+type Estimator interface {
+	getPairProbability(now time.Time, results NodeResults,
+		toNode route.Vertex, amt lnwire.MilliSatoshi,
+		capacity btcutil.Amount) float64
```
For a different ordering of commits, you could introduce the interface, but not yet add `capacity`.

joostjager commented on routing/pathfind_test.go:
```diff
@@ -2756,11 +2775,14 @@ func testProbabilityRouting(t *testing.T, useCache bool,
 		switch {
 		case fromNode == alias["a1"] && toNode == alias["a2"]:
-			return p10
+			return p10 * capacityFactor(amt, capacity)
```
I would be careful with using `capacityFactor` in expectations, because it basically excludes its implementation from the coverage. I know you've covered it separately, but still.

bitromortac replied:
The probability expectations are now hardcoded. This test here may not be that useful as discussed, could remove it.

joostjager replied:
Yes, I'd remove it here.

joostjager reviewed (commented):
Relatively compact pr with a large potential impact. No major comments. Mainly that it might be possible to optimize the commit structure for fewer rewrites. Hold off the addition of the `capacity` parameter until everything is in place. And I think some commits are more readable when squashed.

bitromortac commented on routing/missioncontrol.go:
```diff
@@ -117,7 +118,7 @@ type MissionControl struct {
 	// ProbabilityEstimatorConfig is the config we will use for probability
 	// calculations.
-	ProbabilityEstimatorCfg
+	*ProbabilityEstimatorCfg
```
That was a leftover from the other PR, removed.

bitromortac commented on routing/probability_estimator.go:
```diff
@@ -27,13 +28,24 @@ var (
+	return p.ProbabilityEstimatorCfg
+}
+
+func (p *probabilityEstimator) setConfig(cfg *ProbabilityEstimatorCfg) {
```
Leftover from other PR, removed.

bitromortac commented on routing/pathfind.go:
```diff
@@ -631,7 +632,7 @@ func findPath(g *graphParams, r *RestrictParams, cfg *PathFindingConfig,
 		// Request the success probability for this edge.
 		edgeProbability := r.ProbabilitySource(
-			fromVertex, toNodeDist.node, amountToSend, 0,
+			fromVertex, toNodeDist.node, amountToSend, capacity,
```
I reordered commits such that there should be less rewriting. The only two commits with rewriting are the RPC ones, which I left.

bitromortac commented on routing/unified_policies.go:
```diff
@@ -272,3 +272,48 @@ func (u *unifiedPolicy) minAmt() lnwire.MilliSatoshi {
+// Assumption: we check constraints one more time, but rely on the pathfinder to
+// sort out channels which are unusable. Otherwise this could result in an
+// increased probability later if a zero capacity is returned here.
+func (u *unifiedPolicy) estimateCapacity(
```
I added the test coverage for `getPolicy`.

bitromortac commented on routing/pathfind.go:
```diff
@@ -879,6 +880,8 @@ func findPath(g *graphParams, r *RestrictParams, cfg *PathFindingConfig,
 				amtToSend, g.bandwidthHints,
 			)
 
+			capacity := unifiedPolicy.estimateCapacity(amtToSend)
```
Good point, the methods are very similar. The functionality is now in `getPolicy`, also reduces computation time.

bitromortac commented on routing/graph.go:
```diff
@@ -41,30 +41,29 @@ type CachedGraph struct {
+		return nil, nil, err
+	}
+
+	cleanUp := func() {}
```
Added the comment to the commit message. It's for being able to use the `routingGraph` externally and being able to clean up there. Otherwise this lead to blocking due to the opened db transaction.

joostjager replied:
Wasn't that cleanup possible previously by just calling `Close` on the returned `*CachedGraph` value?

bitromortac commented:
> Relatively compact pr with a large potential impact. No major comments. Mainly that it might be possible to optimize the commit structure for fewer rewrites. Hold off the addition of the capacity parameter until everything is in place. And I think some commits are more readable when squashed.

Thank you for the quick review! I put the capacity calculation into `getPolicy`. Commits are now reordered to have less rewrites, some were squashed.

bitromortac commented on routing/graph.go:
```diff
@@ -86,3 +91,29 @@ func (g *CachedGraph) fetchNodeFeatures(nodePub route.Vertex) (
+		return 0, nil
+	}
+
+	_, capacity := unifiedPolicy.getPolicy(amount, nil)
```
Perhaps it's not ideal to have a `nil` argument here, but it seems a bit too much to initialize empty bandwidth hints for this.

joostjager replied:
Or if `fromNode == g.source` then return 0 right at the top of this method?

joostjager commented on routing/probability.go:
```diff
@@ -0,0 +1,53 @@
+}
+
+// ProbabilityEstimatorCfg contains configuration for our probability estimator.
+type ProbabilityEstimatorCfg struct {
```
Why move this struct? I think it pertains to a specific probability estimator implementation?

joostjager commented on routing/unified_policies.go:
```diff
@@ -195,19 +201,22 @@ func (u *unifiedPolicy) getPolicyLocal(amt lnwire.MilliSatoshi,
 
-	return bestPolicy
+	// We return a zero capacity here to not apply a capacity factor cutoff
+	// for local channels, because we know that the amount should succeed.
```
This sounds like a layering violation. Perhaps clear the capacity higher up?

Or is this an indication that `getPolicy` isn't the right method anymore because it sometimes does and sometimes does not return a capacity?

I know this caused by merging capacity determination into `getPolicy`, but maybe there is a better structure to do a single iteration without dead return values.

joostjager replied:
What could be useful too is this commit: https://github.com/lightningnetwork/lnd/pull/6703/commits/b1bbf3fc42aa60bf6baa174a027394c81a83654e

It allows some extra data to be returned instead of just the db type `CachedEdgePolicy`. I thought that makes sense as we are constructing a sort of synthetic edge that blends many real edges together. Over time, this may evolve towards not exposing the raw `CachedEdgePolicy` at all anymore.

bitromortac replied:
I followed your suggestion and returned a `unifiedPolicyEdge`, although I applied only the minimal change compared to your commit. I think that this really changes the concept of unified policies to unified edges. 

Perhaps we could also do a refactor: `unifiedPolicies` -> `nodeEdgeUnifier`, `unifiedPolicy` -> `edgeUnifier`, and `unifiedPolicyEdge` -> `unifiedEdge`. Also replacing all references to a `policy` with `edge` or `unifiedEdge`.

joostjager replied:
Yes, those renames make sense. I was confused again by the naming when I first looked at this again.

bitromortac replied:
ok, I refactored that, hope it makes more sense now and I didn't change too much

joostjager commented on routing/unified_policies.go:
```diff
@@ -258,7 +279,7 @@ func (u *unifiedPolicy) getPolicyNetwork(
 	modifiedPolicy.TimeLockDelta = maxTimelock
 
-	return &modifiedPolicy
+	return &modifiedPolicy, maxCapMsat.ToSatoshis()
```
Why not stick to msat to avoid rounding max htlc?

bitromortac replied:
As I'm now using `unifiedPolicyEdge` we would have to change it there as well to transport the new units. Also the rounding seems to be not very impactful here

joostjager commented on lnrpc/routerrpc/router_backend.go:
```diff
@@ -108,7 +108,7 @@ type MissionControl interface {
 	// payment from fromNode to toNode.
 	GetProbability(fromNode, toNode route.Vertex,
-		amt lnwire.MilliSatoshi) float64
+		amt lnwire.MilliSatoshi, capacity btcutil.Amount) float64
```
This is a nice commit now. A lot of delta that doesn't need much looking at.

joostjager commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -706,8 +706,14 @@ func (s *Server) QueryProbability(ctx context.Context,
 	mc := s.cfg.RouterBackend.MissionControl
-	// TODO: get node pair capacity
-	prob := mc.GetProbability(fromNode, toNode, amt, 0)
+	capacity, err := s.cfg.RouterBackend.FetchAmountPairCapacity(
```
I would squash this commit and the next with the one that adds `FetchAmountPairCapacity`, to see how it really lands.

joostjager commented on lnrpc/routerrpc/router_backend.go:
```diff
@@ -348,7 +348,10 @@ func (r *RouterBackend) QueryRoutes(ctx context.Context,
 	// that could have been returned from path finding, because mission
 	// control may have been disabled in the provided ProbabilitySource.
-	successProb := r.getSuccessProbability(route)
+	successProb, err := r.getSuccessProbability(route)
```
Longer term, maybe we want to return the success probability directly from `FIndRoute`.

joostjager commented on routing/probability_estimator_test.go:
```diff
@@ -147,17 +153,23 @@ func TestProbabilityEstimatorMix(t *testing.T) {
 
-	ctx.assertPairProbability(testTime, untriedNode, 100, expectedNodeProb)
+	ctx.assertPairProbability(
+		testTime, untriedNode, 100, 0, expectedNodeProb,
```
In a later commit, you're replacing `0` by `testCapacity`. You could just as well do that in this commit and reduce the delta?

bitromortac replied:
I tried to reduce the diff as much as possible. I introduced `testCapacity` early without effect, the question is also whether we should make the probability tests depend explicitly on the `testCapacity` or only inside the assertions?

joostjager replied:
I have no strong preference there. Commit structure looks good now.

joostjager reviewed (commented):
Nice improvement with the commit structure 👍

lightninglabs-deploy commented:
@bitromortac, remember to re-request review from reviewers when ready

joostjager commented on routing/graph.go:
```diff
@@ -86,3 +92,38 @@ func (g *CachedGraph) fetchNodeFeatures(nodePub route.Vertex) (
+		return 0, nil
+	}
+
+	policy := unifiedPolicy.getPolicy(amount, nil)
```
Add comment why no bandwidth hints here? Or can you call `getPolicyNetwork` directly now that it has been ruled out that we are routing from the source node?

joostjager commented on rpcserver.go:
```diff
@@ -695,6 +696,25 @@ func (r *rpcServer) addDeps(s *server, macService *macaroons.Service,
+			}
+			defer func() {
+				closeErr := routingGraph.Close()
+				rpcsLog.Errorf("not able to close routing "+
```
Always log this?

bitromortac replied:
ah right, forgot the nil check :+1:

lightninglabs-deploy commented:
@bitromortac, remember to re-request review from reviewers when ready

positiveblue reviewed (approved):
LGTM @bitromortac

Thank you very much for taking the time of rewriting the commits in an easy way, this is my first time digging into this part of the code and it was pretty easy to follow :1st_place_medal:

joostjager commented on routing/graph.go:
```diff
@@ -86,3 +92,38 @@ func (g *CachedGraph) fetchNodeFeatures(nodePub route.Vertex) (
+func (g *CachedGraph) FetchAmountPairCapacity(nodeFrom, nodeTo route.Vertex,
+	amount lnwire.MilliSatoshi) (btcutil.Amount, error) {
+
+	// For the local node we assume no information on the channel capacity.
```
Why is this necessary? And is it the right point to do this check? From a distance it looks like a pair capacity could still be returned for local channels.

bitromortac replied:
Right, I removed it from the function, for this we already have the local probability calculation and it is a layer violation.

joostjager commented on routing/graph.go:
```diff
@@ -86,3 +92,38 @@ func (g *CachedGraph) fetchNodeFeatures(nodePub route.Vertex) (
+
+	// We may not have all policies available to describe the hop between
+	// the nodes (in the case of hop hints), which is why we return 0 in
+	// this case.
```
I think this comment requires slightly more explanation. Why is it necessary to return 0 in case of hop hints? Maybe the explanation belongs in a different layer, because isn't `CachedGraph` purely about caching the graph without taking into account higher level usage patterns that might involve policies obtained from other sources?

bitromortac replied:
Agree, it is not the right place to have it here. I added a fixup commit with a solution how this could be handled closer to the rpc level, please have a look. I'm not sure if this is the best approach of returning errors here.

bitromortac commented on routing/unified_policies.go:
```diff
@@ -223,6 +224,18 @@ func (u *unifiedPolicy) getPolicyNetwork(
 
+		// Track the maximal capacity for usable channels. If we don't
+		// know the capacity, we fall back to MaxHTLC.
+		capMsat := lnwire.NewMSatFromSatoshis(edge.capacity)
```
this is the case for hop hint channels, yes, as well as for neutrino nodes, which don't have the capacity for channels, as they would have to query each UTXO in the graph

Roasbeef replied:
If a neutrino node is running w/ the default "assume chan valid", then they won't have this information.

Roasbeef commented:
Some linter failures (tho I know we're re-working which ones we actually want to use):
```
^
Error: routing/mock_graph_test.go:2[36](https://github.com/lightningnetwork/lnd/actions/runs/3329622619/jobs/5507074333#step:6:37):3: return with no blank line before (nlreturn)
		return nil
		^
Error: routing/probability_estimator_test.go:265:2: Range statement for test TestCapacityCutoff missing the call to method parallel in test Run (paralleltest)
	for _, test := range tests {
	^
Error: routing/unified_edges_test.go:15:1: Function TestNodeEdgeUnifier missing the call to method parallel (paralleltest)
func TestNodeEdgeUnifier(t *testing.T) {
^
make: *** [Makefile:259: lint] Error 1
Error: Process completed with exit code 2.
```

Roasbeef commented on routing/unified_policies.go:
```diff
@@ -208,6 +208,7 @@ func (u *unifiedPolicy) getPolicyNetwork(
 		bestPolicy  *unifiedPolicyEdge
 		maxFee      lnwire.MilliSatoshi
 		maxTimelock uint16
+		maxCapMsat  lnwire.MilliSatoshi
```
👍 here for seeking out the max size given multiple chans and non strict forwarding and such

Roasbeef commented on routing/unified_policies.go:
```diff
@@ -223,6 +224,18 @@ func (u *unifiedPolicy) getPolicyNetwork(
+			capMsat = edge.policy.MaxHTLC
+		}
+
+		if capMsat > maxCapMsat {
```
Non blocking, but could be a chance to start to introduce some more type params here now that our min version is Go 1.8:
```go
func Max[N constraints.Integer](num1 N, num2 N) N {
     if num1 > num2 {
         return num1
     }
     
     return num2
}
```

In terms of where it should live, we can prob just throw it in `lntypes` for now. In the taro, repo we've started to accumulate some other nice generic channel send/recv timeout funcs: https://github.com/lightninglabs/taro/tree/main/chanutils

Roasbeef commented on routing/unified_policies_test.go:
```diff
@@ -68,26 +74,49 @@ func TestUnifiedPolicies(t *testing.T) {
 	)
+	require.Equal(t, capacity2, policy.capacity)
+
+	// For 4500 msat we expect to get a reduced capacity if we cannot send
```
I think we can improve this test a bit by using `t.Run` here to divide the set of logical assertions in this case, which could very well be individual unit tests in other languages/contexts. 

So we can split it into 5 sub tests here:
  1. 200 msat case 
  2. 400 msat case 
  3. 4500 mst case w/ max htc 
  4. max capacity between two nodes 
  5. no routing info

bitromortac replied:
That makes sense, I added a table-driven test in a first refactor commit and then extended it later when capacity is introduced. The refactor commit for the test could also be squashed into the previous renaming one, but I thought this way it's cleaner.

Roasbeef commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -706,7 +706,7 @@ func (s *Server) QueryProbability(ctx context.Context,
 
 	mc := s.cfg.RouterBackend.MissionControl
-	prob := mc.GetProbability(fromNode, toNode, amt)
+	prob := mc.GetProbability(fromNode, toNode, amt, 0)
```
Perhaps we should actually fetch the capacity information (or have the user supply it) so this is updated to give accurate information?

bitromortac replied:
Right, this is done in a later commit as you noticed

Roasbeef commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -706,7 +706,14 @@ func (s *Server) QueryProbability(ctx context.Context,
 
 	mc := s.cfg.RouterBackend.MissionControl
-	prob := mc.GetProbability(fromNode, toNode, amt, 0)
+	capacity, err := s.cfg.RouterBackend.FetchAmountPairCapacity(
```
👍

Roasbeef commented on routing/probability_estimator_test.go:
```diff
@@ -180,3 +207,70 @@ func TestProbabilityEstimatorMix(t *testing.T) {
+		test := test
+
+		t.Run(test.name, func(t *testing.T) {
+			got := capacityFactor(
```
Can also tack on a `t.Parallel()` here to run each sub-test in parallel.

Roasbeef reviewed (commented):
Great work! Catching up with this and the other PR now, just some minor comments so far. Nice work with the commit structure it made the set of changes pretty easy to follow. Will sit down and play around with some params for the logistic function so I can build up my intuition w.r.t the concrete impact of this in the wild. 

One thing that was somewhat surprising is that we actually didn't need to change any of the existing path finding/ranking tests all that much. Which would mean either: our test coverage wasn't that great before, the impact on edge selection is negligible, or the test cases we had weren't set up to insert cases where the old vs the new probability estimation would differ drastically.

yyforyongyu commented on routing/probability_estimator.go:
```diff
@@ -10,6 +10,22 @@ import (
+	// probability may be hit because of channel exhaustion and with only
+	// the apriori probability known i.e.,
+	// P_apriori^3 * P_cutoff_min > P_min, where P_cutoff_min ~ 0.075.
+	capacityCutoffFraction   = 0.75
```
and why are these two figures chosen?

bitromortac replied:
I'm not 100% sure we should make those configurable as the parameters interact in some way as given in the updated description, which should make the choice of parameters more clear. Could expose them if there's strong demand though!

yyforyongyu commented on routing/probability_estimator_test.go:
```diff
@@ -104,12 +118,25 @@ func TestProbabilityEstimatorOneSuccess(t *testing.T) {
 		testTime, node1, 100, testCapacity, aprioriPrevSucProb,
 	)
 
+	// A high success probability is not influcenced by the capacity factor.
```
Not sure I understand what's been tested here.

bitromortac replied:
I updated the comment, hope it's more clearer now

yyforyongyu commented on routing/probability_estimator_test.go:
```diff
@@ -180,3 +207,70 @@ func TestProbabilityEstimatorMix(t *testing.T) {
+
+// TestCapacityCutoff tests the mathematical expression and limits for the
+// capacity factor.
+func TestCapacityCutoff(t *testing.T) {
```
👍

yyforyongyu commented on routing/probability_estimator_test.go:
```diff
@@ -180,3 +207,70 @@ func TestProbabilityEstimatorMix(t *testing.T) {
+		{
+			// Even when we spend the full capacity, we still want
+			// to have some residual to not throw away routes due to
+			// minimal probability constraints.
```
what is this "minimal probability constraints"?

bitromortac replied:
I updated the comment and explained this a bit more in the comment about the parameters. Currently, we have a requirement for a minimal probability for routes. If we dip below that we'd not consider the route.

yyforyongyu commented on routing/router.go:
```diff
@@ -2798,13 +2797,13 @@ func (r *ChannelRouter) BuildRoute(amt *lnwire.MilliSatoshi,
 		// Add fee for this hop.
 		if !localChan {
-			runningAmt += policy.policy.ComputeFee(runningAmt)
+			runningAmt += edge.policy.ComputeFee(runningAmt)
```
hahah I see you already doing this, dope!

yyforyongyu commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -705,14 +705,21 @@ func (s *Server) QueryProbability(ctx context.Context,
+	// constraints, so we return a zero probability.
 	if err != nil {
-		return nil, err
+		log.Debugf("Cannot fetch capacity: %v", err)
```
nit: this could use a `log.Errorf`

yyforyongyu reviewed (changes requested):
Very nice PR! I think we all agree here that the PR is really reviewer-friendly! Awesome job! No major comments, just a few nits plus some questions, and the linter needs to be fixed.

PS, I think it'd be dope to have PR that creates a doc detailing the path-finding algorithm used in `lnd`.

bitromortac commented on routing/router.go:
```diff
@@ -2798,10 +2798,11 @@ func (r *ChannelRouter) BuildRoute(amt *lnwire.MilliSatoshi,
 		// Add fee for this hop.
 		if !localChan {
-			runningAmt += policy.ComputeFee(runningAmt)
+			runningAmt += policy.policy.ComputeFee(runningAmt)
```
right, there's the rename-refactor as you saw later :slightly_smiling_face:

bitromortac commented:
Thank you all for the reviews and sorry for the delay!

> Great work! Catching up with this and the other PR now, just some minor comments so far. Nice work with the commit structure it made the set of changes pretty easy to follow. Will sit down and play around with some params for the logistic function so I can build up my intuition w.r.t the concrete impact of this in the wild.

Thank you @roasbeef, I added some more explanations to the comment where the params are defined.

> One thing that was somewhat surprising is that we actually didn't need to change any of the existing path finding/ranking tests all that much. Which would mean either: our test coverage wasn't that great before, the impact on edge selection is negligible, or the test cases we had weren't set up to insert cases where the old vs the new probability estimation would differ drastically.

Right, there were no test cases where the channel capacity was used completely, but only tiny fractions such that the capacity factor would not really be active. I think it shows that the change is conservative for small amounts. The effect of this change would have been greater if the accuracy in `assertPairProbability` had been higher.

> Very nice PR! I think we all agree here that the PR is really reviewer-friendly! Awesome job! No major comments, just a few nits plus some questions, and the linter needs to be fixed.
> 
> PS, I think it'd be dope to have PR that creates a doc detailing the path-finding algorithm used in lnd.

Thanks for the nice words @yyforyongyu! I fully agree, it would not be out of scope for this PR. Perhaps such a doc would make even more sense after we've decided about #6815.

> Thank you very much for taking the time of rewriting the commits in an easy way, this is my first time digging into this part of the code and it was pretty easy to follow 1st_place_medal

Cool @positiveblue, this is also due to Joost's nice suggestions.

I rebased on master to fix the linter issues. From my side, it would be great to know if the changes in the RPCs make sense, because there's an error path that has opened for `QueryRoutes` if no capacity could be determined for a channel along the route. `QueryProb` behaves different currently, it returns a zero probability if there's missing info. Because there's some unclarity about the RPCs I added a temporary commit to fix an itest, would squash that later.

joostjager commented on lnrpc/routerrpc/router_backend.go:
```diff
@@ -343,7 +348,12 @@ func (r *RouterBackend) QueryRoutes(ctx context.Context,
 	// that could have been returned from path finding, because mission
 	// control may have been disabled in the provided ProbabilitySource.
-	successProb := r.getSuccessProbability(route)
+	successProb, err := r.getSuccessProbability(
```
Still wondering if with the newly added extra code and lookup it isn't better to return the probably along with the route from pathfinding. Pathfinding already has it.

bitromortac replied:
I have added a commit that returns the probability from `findPath` to be usable in `QueryRoutes`. Advantage is that we don't have to recompute the probability (see fixup commit that removes `getSuccessProbability`), but the downside is that we don't get probability estimates anymore if `use_mc` is disabled. Unfortunately we can't remove the capacity lookup code because `QueryProbability` still needs it.

joostjager replied:
I think it's a good change. Probabilities without `use_mc` doesn't make that much sense anyway.

joostjager commented on docs/release-notes/release-notes-0.16.0.md:
```diff
@@ -191,6 +191,11 @@ certain large transactions](https://github.com/lightningnetwork/lnd/pull/7100).
   using the address in question](
   https://github.com/lightningnetwork/lnd/pull/7025)
 
+## Pathfinding
```
I usually squash this with the commit that contains the actual beef.

bitromortac replied:
makes sense, done!

joostjager commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -709,8 +709,22 @@ func (s *Server) QueryProbability(ctx context.Context,
+	} else {
+		prob = mc.GetProbability(fromNode, toNode, amt, capacity)
+	}
+
```
More generally I am wondering what this `QueryProbability` is actually used for in the real world and whether it is worth the maintenance that it requires. This is just a side comment.

joostjager replied:
What would you think of just marking it as deprecated in 0.16, and then remove the rpc plus `FetchAmountPairCapacity` in 0.17?

Perhaps `QueryMissionControlRequest` can be extended with from/to fields to return the base data for the probability calculation. If someone really needs a probability, they can always link against lnd "the library" and calculate it client-side.

bitromortac replied:
> I am wondering what this QueryProbability is actually used for in the real world

Wondering about this too, it looks like right now it's only used for debugging. I added a commit to mark it deprecated in the rpc and lncli, hope I found all the places.

joostjager reviewed (approved):
Nothing big from my side. LGTM.

joostjager commented on routing/router.go:
```diff
@@ -1809,7 +1808,7 @@ func (r *ChannelRouter) FindRoute(source, target route.Vertex,
 	)
 
-	return route, nil
+	return route, probability, nil
```
I am not sure about that. Routes can exist without a probability. They are a foundational data structure in lightning, where as probability is just a field that a specific pathfinding implementation adds to it.

bitromortac replied:
Understand both points, would be grateful for more input.

bitromortac commented on routing/unified_edges.go:
```diff
@@ -222,6 +223,14 @@ func (u *edgeUnifier) getEdgeNetwork(amt lnwire.MilliSatoshi) *unifiedEdge {
+		// know the capacity, we fall back to MaxHTLC.
+		capMsat := lnwire.NewMSatFromSatoshis(edge.capacity)
+		if capMsat == 0 && edge.policy.MessageFlags.HasMaxHtlc() {
+			capMsat = edge.policy.MaxHTLC
```
I think we would only like to do this if we are in `assumeChannelValid=false` (normal) mode, otherwise this could become very spammy - every channel would fall back to `MaxHTLC`. I could add a `Trace` statement here. Alternatively we have to pass in `assumeChannelValid`, which would perhaps be too much just for logging?

yyforyongyu replied:
yeah I think this only happens in neutrino. Using a `Trace` sounds good.

yyforyongyu commented on lnrpc/routerrpc/router_backend.go:
```diff
@@ -330,7 +330,7 @@ func (r *RouterBackend) QueryRoutes(ctx context.Context,
 	// can carry `in.Amt` satoshis _including_ the total fee required on
 	// the route.
-	route, _, err := r.FindRoute(
+	route, successProb, err := r.FindRoute(
```
The `fixup` commit needs to be squashed.

bitromortac replied:
Was not sure if I had to revert it, but it's squashed now.

yyforyongyu commented on routing/probability_estimator.go:
```diff
@@ -10,6 +10,43 @@ import (
 )
 
+const (
+	// capacityCutoffFraction and capacitySmearingFraction define how
```
👍

yyforyongyu commented on cmd/lncli/cmd_mission_control.go:
```diff
@@ -163,7 +163,7 @@ func queryMissionControl(ctx *cli.Context) error {
 	Name:      "queryprob",
 	Category:  "Mission Control",
-	Usage:     "Estimate a success probability.",
+	Usage:     "Deprecated. Estimate a success probability.",
```
nit: For cli commands we can add the `Hidden` param to hide it.

bitromortac replied:
Cool, done!

yyforyongyu commented on lnrpc/routerrpc/router_backend.go:
```diff
@@ -103,7 +103,7 @@ type MissionControl interface {
 	// payment from fromNode to toNode.
 	GetProbability(fromNode, toNode route.Vertex,
-		amt lnwire.MilliSatoshi) float64
+		amt lnwire.MilliSatoshi, capacity btcutil.Amount) float64
```
This is probably an edgy case, but what if the channel capacity is actually zero?

bitromortac replied:
Good question, is it even possible to have a channel with zero capacity, or do you refer to an error in capacity validation? In that case the capacity factor would not be active (it would be 1), so no change compared to the current situation. Either the channel has a `maxHTLCMsat` set, then it must be less than or equal to the capacity per spec. In that scenario it would be caught by the `amtInRange` check. The `amtInRange` check currently doesn't prohibit sending to a zero sat channel:
https://github.com/lightningnetwork/lnd/blob/e23c5dca7364bcd329559d86727d5e2e4a8bbcc3/routing/unified_policies.go#L104

In the worst-case scenario we would send to that route and would fail, avoiding the channel next time.

lightninglabs-deploy commented:
@roasbeef: review reminder
@bitromortac, remember to re-request review from reviewers when ready

Roasbeef commented:
I think this is ready to land after a rebase!
